jenkins_job_manager/raw_ext.py

```diff
@@ -0,0 +1,81 @@
+"""
+Contains extensions for raw xml loading with jjb
+"""
+import os
+import logging
+import jinja2
+import xml.etree.ElementTree
+import jenkins_jobs.modules.base
+from jenkins_jobs.xml_config import XmlJob, XmlJobGenerator
+
+log = logging.getLogger("jjm")
+
+
+def xml_escape(s):
+    """hacks ElementTree to give us strings escaped for xml"""
+    ET = xml.etree.ElementTree
+    a = ET.Element("a")
+    a.text = s
+    # spit out element as string, slice out the element text
+    escaped_s = ET.tostring(a, encoding="unicode")[3:-4]
+    return escaped_s
+
+
+def load_xml_escaped(path):
+    with open(path) as fp:
+        contents = fp.read()
+    return xml_escape(contents)
+
+
+# initialize a global jinja env
+jenv = jinja2.Environment(
+    loader=jinja2.FileSystemLoader(searchpath=(os.curdir,)),
+    undefined=jinja2.StrictUndefined,
+)
+jenv.globals["load_xml_escaped"] = load_xml_escaped
+jenv.filters["xml_escape"] = xml_escape
+
+
+class RawXmlProject(jenkins_jobs.modules.base.Base):
+    """add a job type for raw xml"""
+
+    jenv = jenv
+
+    def root_xml(self, data):
+        name = data["name"]
+        xmlstr = data["raw"]
+        # add in a hack for raw jobs to do optional jinja templating
+        if data.get("jinja") is True:
+            log.debug("Rendering jinja template for %s", name)
+            templ = self.jenv.from_string(xmlstr)
+            xmlstr = templ.render(data=data)
+            log.debug("rendered: \n---\n%s\n---", xmlstr)
+        xml_parent = xml.etree.ElementTree.fromstring(xmlstr)
```

Comment:
I know this is likely a copy/paste from the old file, but:
Using xml.etree.ElementTree.fromstring to parse untrusted XML data is known to be vulnerable to XML attacks. Replace xml.etree.ElementTree.fromstring with its defusedxml equivalent function or make sure defusedxml.defuse_stdlib() is called

Reply:
These are local files. A xml injection is highly unlikely.

Reply:
![tumblr_liwcmw6Kl51qg4umeo1_500](https://user-images.githubusercontent.com/10673023/103389034-b7f37600-4ac9-11eb-9139-3505464773fb.gif)

IRL i trust though